entity-framework/core/miscellaneous/events/interception.md

```diff
@@ -0,0 +1,750 @@
+
+Often, the trickiest part of the interception is determining when the command corresponds to the query that needs to be modified. Parsing the SQL is one option, but tends to be fragile. Another option is to use [EF Core query tags](xref:core/querying/tags) to tag each query that should be modified. For example:
+
+<!--
+            var blogs1 = context.Blogs.TagWith("Use hint: robust plan").ToList();
+-->
+[!code-csharp[TaggedQuery](../../../../samples/core/Miscellaneous/CommandInterception/Program.cs?name=TaggedQuery)]
+
+This tag can then be detected in the interceptor as it will always be included as a comment in the first line of the command text. On detecting the tag, the query SQL is modified to add the appropriate hint:
+
+<!--
+public class TaggedQueryCommandInterceptor : DbCommandInterceptor
+{
+    public override InterceptionResult<DbDataReader> ReaderExecuting(
+        DbCommand command,
+        CommandEventData eventData,
+        InterceptionResult<DbDataReader> result)
+    {
+        ManipulateCommand(command);
+
+        return result;
+    }
+
+    public override ValueTask<InterceptionResult<DbDataReader>> ReaderExecutingAsync(
+        DbCommand command,
+        CommandEventData eventData,
+        InterceptionResult<DbDataReader> result,
+        CancellationToken cancellationToken = default)
+    {
+        ManipulateCommand(command);
+
+        return new ValueTask<InterceptionResult<DbDataReader>>(result);
+    }
+
+    private static void ManipulateCommand(DbCommand command)
+    {
+        if (command.CommandText.StartsWith("-- Use hint: robust plan", StringComparison.Ordinal))
+        {
+            command.CommandText += " OPTION (ROBUST PLAN)";
+        }
+    }
+}
+-->
+[!code-csharp[TaggedQueryCommandInterceptor](../../../../samples/core/Miscellaneous/CommandInterception/TaggedQueryCommandInterceptor.cs?name=TaggedQueryCommandInterceptor)]
+
+Notice:
+
+* The interceptor inherits from <xref:Microsoft.EntityFrameworkCore.Diagnostics.DbCommandInterceptor> to avoid having to implement every method in the interceptor interface.
+* The interceptor implements both sync and async methods. This ensures that the same query hint is applied to sync and async queries.
+* The interceptor implements the `Executing` methods which are called by EF Core with the generated SQL _before_ it is sent to the database. Contrast this with the `Executed` methods, which are called after the database call has returned.
+
+Running the code in this example generates the following when a query is tagged:
+
+```sql
+-- Use hint: robust plan
+
+SELECT [b].[Id], [b].[Name]
+FROM [Blogs] AS [b] OPTION (ROBUST PLAN)
+```
+
+On the other hand, when a query is not tagged, then it is sent to the database unmodified:
+
+```sql
+SELECT [b].[Id], [b].[Name]
+FROM [Blogs] AS [b]
+```
+
+### Example: Connection interception for SQL Azure authentication using ADD
```

Comment:
This is obviously useful for people, but I'm worried we're showing a perf anti-pattern here (basically what I wrote [here](https://github.com/dotnet/efcore/issues/13124#issuecomment-709901961)).

Is it really considered good practice to fetch AAD access tokens *every time* a connection is opened? Isn't it possible to cache these for at least a while?

Reply:
Confirmed with Azure.Identity that perf-wise this is problematic, see https://github.com/dotnet/efcore/issues/13261#issuecomment-714923330